rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala

```diff
@@ -0,0 +1,239 @@
+package coop.rchain.rholang.interpreter
+
+import Env.DeBruijn
+
+sealed trait Substitute
+
+object Substitute {
+
+  def substitute(term: Par, env: Env): Par =
+    Par(
+      for { send <- term.sends } yield {
+        substitute(send, env)
+      },
+      for { recv <- term.receives } yield {
+        substitute(recv, env)
+      },
+      for { eval <- term.evals } yield {
+        substitute(eval, env)
+      },
+      for { neu <- term.news } yield {
+        substitute(neu, env)
+      },
+      term.exprs, /*for { exp <- term.exprs } yield {
+        case e @ EVar(_) =>
+          substitute(e, env) match {
+            case Left(varue) => EVar(varue)
+            case Right(Left(quote)) =>
+              throw new Error(s"Channel must be term in process expression: $quote")
+            case Right(Right(par)) => return term.merge(par)
+          }
+        case _ => substitute(exp, env)
+      },*/
+      term.id
+    )
+
+  def rename(term: Par, j: Int): Par =
+    Par(
+      for (send <- term.sends) yield {
+        rename(send, j)
+      },
+      for (recv <- term.receives) yield {
+        rename(recv, j)
+      },
+      for (eval <- term.evals) yield {
+        rename(eval, j)
+      },
+      for (neu <- term.news) yield {
+        rename(neu, j)
+      },
+      for (exp <- term.exprs) yield {
+        rename(exp, j)
+      },
+      term.id
+    )
+
+  def substitute(term: Quote, env: Env): Quote = Quote(substitute(term.p, env))
+
+  def rename(term: Quote, j: Int): Quote = Quote(rename(term.p, j))
+
+  def substitute(term: Channel, env: Env): Either[Channel, Par] =
+    term match {
+      case q @ Quote(_) => Left(substitute(q, env))
+      case ChanVar(b @ _) =>
+        substitute(b, env) match {
+          case Right(Left(quote)) => Left(quote)
+          case Right(Right(par))  => Right(par)
+          case Left(c @ _)        => Left(ChanVar(c))
+        }
+    }
+
+  def rename(term: Channel, j: Int): Channel =
+    term match {
+      case q @ Quote(_)   => rename(q, j)
+      case ChanVar(varue) => ChanVar(rename(varue, j))
+    }
+
+  def substitute(term: Send, env: Env): Send =
+    substitute(term.chan, env) match {
```

Comment:
substituting into a channel should only ever yield a channel. It shouldn't yield an either.